src/coreComponents/constitutive/fluid/MultiFluidBase.hpp

```diff
@@ -339,6 +478,322 @@ class MultiFluidBase : public ConstitutiveBase
 
 };
 
+template< typename OUT_ARRAY >
+GEOSX_HOST_DEVICE
+void
+MultiFluidBase::KernelWrapper::
+  convertToMoleFractions( arraySlice1d< real64 const, compflow::USD_COMP - 1 > const composition,
+                          arrayView1d< real64 const > const componentMolarWeight,
+                          OUT_ARRAY && compMoleFrac ) const
+{
+  integer const numComps = numComponents();
+
+  real64 totalMolality = 0.0;
+  for( integer ic = 0; ic < numComps; ++ic )
+  {
+    compMoleFrac[ic] = composition[ic] / componentMolarWeight[ic]; // this is molality (units of mole/mass)
+    totalMolality += compMoleFrac[ic];
+  }
+
+  real64 const totalMolalityInv = 1.0 / totalMolality;
+  for( integer ic = 0; ic < numComps; ++ic )
+  {
+    compMoleFrac[ic] *= totalMolalityInv;
+  }
+}
+
+template< integer maxNumComp, typename OUT_ARRAY >
+GEOSX_HOST_DEVICE
+void
+MultiFluidBase::KernelWrapper::
+  convertToMoleFractions( arraySlice1d< real64 const, compflow::USD_COMP - 1 > const composition,
+                          arrayView1d< real64 const > const componentMolarWeight,
+                          OUT_ARRAY && compMoleFrac,
+                          real64 (& dCompMoleFrac_dCompMassFrac)[maxNumComp][maxNumComp] ) const
+{
+  integer const numComps = numComponents();
+
+  real64 totalMolality = 0.0;
+  for( integer ic = 0; ic < numComps; ++ic )
+  {
+    real64 const mwInv = 1.0 / componentMolarWeight[ic];
+    compMoleFrac[ic] = composition[ic] * mwInv; // this is molality (units of mole/mass)
+    dCompMoleFrac_dCompMassFrac[ic][ic] = mwInv;
+    totalMolality += compMoleFrac[ic];
+  }
+
+  real64 const totalMolalityInv = 1.0 / totalMolality;
+  for( integer ic = 0; ic < numComps; ++ic )
+  {
+    compMoleFrac[ic] *= totalMolalityInv;
+
+    for( integer jc = 0; jc < numComps; ++jc )
+    {
+      dCompMoleFrac_dCompMassFrac[ic][jc] -= compMoleFrac[ic] / componentMolarWeight[jc];
+      dCompMoleFrac_dCompMassFrac[ic][jc] *= totalMolalityInv;
+    }
+  }
+}
+
+template< integer maxNumPhase >
+GEOSX_HOST_DEVICE
+void
+MultiFluidBase::KernelWrapper::
+  convertToMassFractions( real64 const (&phaseMolecularWeight)[maxNumPhase],
+                          arraySlice1d< real64, multifluid::USD_PHASE - 2 > const phaseFrac,
+                          arraySlice2d< real64, multifluid::USD_PHASE_COMP - 2 > const phaseCompFrac ) const
+{
+  integer const numPhase = numPhases();
+  integer const numComp = numComponents();
+
+  // 1. Convert phase fractions (requires two passes)
+  real64 totalMass{};
+
+  // 1.1. Compute mass of each phase and total mass (on a 1-mole basis)
+  for( integer ip = 0; ip < numPhase; ++ip )
+  {
+
+    bool const phaseExists = (phaseFrac[ip] > 0);
+    if( !phaseExists )
+    {
+      continue;
+    }
+
+    phaseFrac[ip] *= phaseMolecularWeight[ip];
+    totalMass += phaseFrac[ip];
+  }
```

Comment:
For your first question, you are right, the fractions for non-existent phases are just zero, so we can just skip them here. Specifically, the places where we set the phase fractions to zero for non-existent phases are:
- Black-Oil: https://github.com/GEOSX/GEOSX/blob/343d427a3eef852a50e7a73a372a0b9df0c24355/src/coreComponents/constitutive/fluid/BlackOilFluid.hpp#L524
- CO2-brine: https://github.com/GEOSX/GEOSX/blob/343d427a3eef852a50e7a73a372a0b9df0c24355/src/coreComponents/constitutive/fluid/PVTFunctions/CO2Solubility.hpp#L178
- Compositional: https://github.com/GEOSX/PVTPackage/blob/064526518b966fea1ae6f722a0903e934827ef4a/PVTPackage/source/MultiphaseSystem/PhaseSplitModel/CompositionalFlash.cpp#L99

Therefore, I removed these checks in the latest commits.

For your second question, yes I agree, I think it would be nice to have the `forExistingPhases` lambda in the compositional solver kernels where we divide by density. Much better than the `continue` keywords that I put in a few places. I have not implemented this in this PR to avoid touching other files, but it would be good for a future refactoring.